### ISSUE TITLE
Add ability to calculate the median expected discovery significance

### ISSUE THREAD

**Issue by matthewfeickert** [scikit-hep/pyhf#1810]
### Summary

In PR #1232 we added the ability to calculate the discovery test statistic. However, we also force the parameter of interest to be `0` (we were going for the fact that this is the **discovery** test statistic).

https://github.com/scikit-hep/pyhf/blob/e4331091a123615ecc91bf9be988d799368c705d/src/pyhf/infer/test_statistics.py#L480-L484 

This is a reasonable default, but if you want to calculate the median expected significance in the presence of a signal you would need to calculate `q0(mu=1)`. This is currently not allowed with the `v0.6.3` API.

It would be good if we provided a way to calculate the median expected significance, but kept the warning and required an additional boolean be added. E.g.

```python
q0(mu, data, pdf, init_pars, par_bounds, fixed_params, return_fitted_pars=False, median=False):
...
    if mu != 0.0 and median != True:
        log.warning(
            "If not calculating the expected median significance, the q0 test statistic is only used for fit configurations with POI set to zero. Setting mu=0."
            + "\n"
            + "To calculate the expected median significance, pass kwarg median=True to pyhf.infer.test_statistics.q0()."
        )
        mu = 0.0
```

### Additional Information

I was reminded of this today when reading through [Eilam Gross's lectures from the 2017 European School of High-Energy Physics](https://indico.cern.ch/event/614672/), in particular Slide 79 "The Median Sensitivity (via ASIMOV)".

### Code of Conduct

- [X] I agree to follow the Code of Conduct

**Comment by matthewfeickert** [comment]
I should note of course, that you can do something similar to like what @alexander-held is doing now in `cabinetry` with [`cabinetry.fit.significance`](https://cabinetry.readthedocs.io/en/stable/api.html#cabinetry.fit.significance)

```python
import pyhf
from scipy.stats import norm

mu = 1.0
model = pyhf.simplemodels.uncorrelated_background([6], [9], [3])
data = [9] + model.config.auxdata
init_pars = model.config.suggested_init()
par_bounds = model.config.suggested_bounds()
fixed_params = model.config.suggested_fixed()

obs_p_val, exp_p_val = pyhf.infer.hypotest(
    0.0,
    data,
    model,
    init_pars=init_pars,
    fixed_params=fixed_params,
    par_bounds=par_bounds,
    test_stat="q0",
    return_expected=True,
)
obs_p_val = float(obs_p_val)
exp_p_val = float(exp_p_val)
obs_significance = norm.isf(obs_p_val, 0, 1)
exp_significance = norm.isf(exp_p_val, 0, 1)

print(f"Observed significance: {obs_significance}")
print(f"Expected significance: {exp_significance}")
```
```
Observed significance: 0.0
Expected significance: 1.358485746768557
```

but this still faces the same problems if someone has a workspace in which they have fixed the POI (they might have intentionally fixed it to `1` or something to blind the workspace. `WSMaker` does this for example by default.)

**Comment by matthewfeickert** [comment]
This is motivated by trying to work on a translation recipe for `WSMaker` in https://github.com/scikit-hep/pyhf/issues/1341

**Comment by matthewfeickert** [comment]
Yeah, I think just stealing what `cabinetry` is doing and then also implementing Issue #1712 would be the way to go here.

**Comment by alexander-held** [comment]
I'm not sure whether I understand the direction of this issue correctly. Is it specifically about calculating expected discovery significance without any reference to observed data? `cabinetry.fit.significance` does not do that, it merely uses the existing `pyhf` API to get both observed + expected discovery significance. The expected discovery significance is evaluated with Asimov data built from a fit to observed data with mu=1 fixed. This already exists within `pyhf.infer.hypotest`, and the value of 1 is fixed:
https://github.com/scikit-hep/pyhf/blob/fc3cc62a44236002d68d5606bf213105fc07b088/src/pyhf/infer/calculators.py#L378

(There may be some value in allowing other values here, but users could also re-scale their signal instead.)

If this is about doing a fully blind evaluation ("pre-fit Asimov") of the expected discovery significance, then the way to do this in cabinetry would be via `cabinetry.model_utils.asimov_data` (there are different possibilities for the treatment of free-floating parameters, `cabinetry` uses the initial parameter settings as nominal values). This works also with pure `pyhf` (depending on what exactly is desired as a dataset) by passing a suitable `data` to `hypotest`.

> but this still faces the same problems if someone has a workspace in which they have fixed the POI (e.g. WSMaker intentionally fixes the POI to 1 or something to blind the workspace.) This is a more general problem with workspace conversion though (c.f. Issue https://github.com/scikit-hep/pyhf/issues/1341) that isn't really part of this Issue.

I see two options here, either ignoring what is set in the workspace for convenience (so that the calculation can go forward), or raising an error with a suitable message. I see value in both approaches.
